### PR TITLE
Describe use of the `Instance` trait type in the user manual

### LINKED ISSUE #1179: Describe use of the `Instance` trait type in the user manual
The `Instance` trait type is one of the most fundamental and useful after the scalar types, but is barely covered in the user guide.

It would be useful to have something in the user manual. `defining.rst` might be a good place to start.

### PR DESCRIPTION
**Pull request by aaronayres35**
Closes #1179 
Related to #1214 (not sure if we want to do more to warn about / discourage this)

This PR adds a brief section to the documentation giving more detail about the predefined `Instance` trait type.

**Checklist**
- ~[ ] Tests~ N/A
- ~[ ] Update API reference (`docs/source/traits_api_reference`)~ N/A
- [x] Update User manual (`docs/source/traits_user_manual`)
- ~[ ] Update type annotation hints in `traits-stubs`~ N/A

### CODE REVIEW

File: docs/source/traits_user_manual/defining.rst [hunk 1]
```diff
@@ -405,6 +405,57 @@ the table.
...
+    class Parent(HasTraits):
+        pass
+
+    class Child(HasTraits):
+        # default value is None
+        father = Instance(Parent())
+        # default value is also None
+        grandfather = Instance(Parent)
+        # default value is Parent()
+        mother = Instance(Parent, args=())
```
Comment (aaronayres35) [inline docs/source/traits_user_manual/defining.rst:439]: Maybe this should be changed to better mirror this format, as if it were being typed in a python shell?

Screen Shot 2021-01-14 at 12 38 27 PM

Comment (aaronayres35) [reply docs/source/traits_user_manual/defining.rst:439]: There is also #1274 open currently.  I sort of blindly followed the existing pattern by creating a instance_trait_defaults.py and copying code here, but maybe a literal include is better.  If doing the python shell style though then thats different.

File: docs/source/traits_user_manual/defining.rst [hunk 2]
```diff
@@ -405,6 +405,57 @@ the table.
 |                  | *allow_none* = False, *adapt* = 'yes', \*\*\ *metadata*])|
 +------------------+----------------------------------------------------------+
 
+.. index:: Instance trait
+
+.. _instance:
+
+Instance
+::::::::
+One of the most fundamental and useful predefined trait types is Instance.
```
Comment (mdickinson) [inline docs/source/traits_user_manual/defining.rst:414]: Can we make this a link to the API definition? (Something like ```:class:`~.Instance` ``` should work.)

File: docs/source/traits_user_manual/defining.rst [hunk 3]
```diff
@@ -405,6 +405,57 @@ the table.
...
+Instance trait values are an instance of a particular class or its subclasses,
+as specified by the **klass** argument. **klass** can be either an instance of
+a class or a class itself (note this applies to all python classes, not
+necessarily just ``HasTraits`` subclasses). Further, **klass** can also be an
```
Comment (mdickinson) [inline docs/source/traits_user_manual/defining.rst:418]: I think we should drop the sentence about interfaces. Historical background: `Instance(ISomeInterface)` checks used to work differently from `Instance(SomeRandomClass)` checks - the former genuinely would check that the type of the assigned value had all the right methods. That's most of what the stuff in the "interfaces" section is discussing. (And yes, that section needs to be updated and revised, but one thing at a time ...)

So it's true that `klass` can be an interface, but in that case, in modern Traits, the check is still an `isinstance(assigned_object, ISomeInterface)`, just like it is for regular classes.

It's tricky, because use of interfaces here _is_ an important use-case, but I don't want to give the impression that interfaces are somehow treated differently from other classes - we're just using a simple `isinstance` check in all cases.

File: docs/source/traits_user_manual/defining.rst (hunk 1 above)
Comment (mdickinson) [reply docs/source/traits_user_manual/defining.rst:439]: I think I find it clearer with the class definitions as they are (so not in an interactive context). But you can always mix and match: show the class definitions in one block, and then follow up with something that looks like an interactive session using those class definitions.

File: docs/source/traits_user_manual/defining.rst [hunk 4]
```diff
@@ -405,6 +405,57 @@ the table.
...
+created, but rather is lazily instantiated when the trait is first accessed.
+The default ``Parent`` will also be instantiated if the trait is assigned to
+and there is a change handler defined on the trait (to detect changes from the
+default value). For more details on change handlers and trait notification see
+:ref:`observe-notification`.
+
+Somewhat surprisingly, ``mother = Instance(Parent, ())`` will also yield a
+default value of ``Parent()``, even though in that case it is **factory** that
+is ``()`` not **args**.  This is a result of implementation details, and using
+this form is discouraged.  Another discouraged behavior is passing in another
```
Comment (mdickinson) [inline docs/source/traits_user_manual/defining.rst:454]: On `factory` being `()`, we should make it clear what the preferred way of writing the code is.  The `Instance(SomeClass, ())` is a _very_ common idiom in existing code.

Comment (mdickinson) [reply docs/source/traits_user_manual/defining.rst:454]: For the second, maybe change "discouraged behaviour" to "common mistake"?

**Review by mdickinson — COMMENTED** [review]
This is looking good; thank you! I've left a few comments.

File: docs/source/traits_user_manual/defining.rst (hunk 4 above)
Comment (aaronayres35) [reply docs/source/traits_user_manual/defining.rst:454]: > On factory being (), we should make it clear what the preferred way of writing the code is. The Instance(SomeClass, ()) is a very common idiom in existing code.

Just for my own understanding, the reason we are advising against this is just it is less clear/explicit and might not be doing what you think it is under the hood, correct?  Basically Instance is doing some DWIM stuff and it sees factory is a tuple `()` so it sets factory to klass and args to ().  So it effectively does exactly the same thing as `Instance(SomeClass, args=())`, but sort of secretly.

In other words my real question is there isn't any actual difference in end behavior correct? / there aren't any strange and rare bugs this directly causes?

Comment (aaronayres35) [reply docs/source/traits_user_manual/defining.rst]: > For the second, maybe change "discouraged behaviour" to "common mistake"?

@mdickinson is this section added in the docs enough to close #1214.  The issue mentions that we might want to consider even making it an error if we can't think of any valid uses.  We also may want to explicitly add a warning in the code not just the documentation?

Comment (mdickinson) [reply docs/source/traits_user_manual/defining.rst]: > Just for my own understanding, the reason we are advising against this is [...]

Um, actually, I don't know. Where did you get the information that we're advising against this?!

Honestly, I don't think we can or should advise against - the pattern is so common that we should probably just keep it, and I'm not aware of any actual problems that creates.

I _would_ still *recommend* passing the args by keyword instead of position (so `Instance(SomeClass, args=())` rather than `Instance(SomeClass, ())`), just for the benefit of being explicit.

But I think that's different from actually discouraging the second form. I don't want to users to get the impression that the second form is likely to be deprecated.

Comment (mdickinson) [reply docs/source/traits_user_manual/defining.rst]: > In other words my real question is there isn't any actual difference in end behavior correct?

There's nothing that I'm aware of.

> is this section added in the docs enough to close #1214

Hmm. Good question. Probably not. I'm somewhat inclined to close #1214 anyway, though - warning about `Instance(Int)` seems like something that would be more reasonable in a static checker than in Traits itself.

Comment (aaronayres35) [reply docs/source/traits_user_manual/defining.rst]: > Um, actually, I don't know. Where did you get the information that we're advising against this?!

In issue #1180 the comment "Also explicitly document (or disparage) the equivalent form Instance(SomeClass, ())".  Sounds like we are more on the document rather than disparage mindset.

> Honestly, I don't think we can or should advise against - the pattern is so common that we should probably just keep it, and I'm not aware of any actual problems that creates.
> 
> I _would_ still _recommend_ passing the args by keyword instead of position (so `Instance(SomeClass, args=())` rather than `Instance(SomeClass, ())`), just for the benefit of being explicit.
> 
> But I think that's different from actually discouraging the second form. I don't want to users to get the impression that the second form is likely to be deprecated.

That make sense, I can update the relevant sentence to not say "using this form is discouraged", and just say something like "This is a result of implementation details, however the recommended way of writing this code is to explicitly pass **args** by keyword like so `mother = Instance(Parent, args=())`."

Comment (mdickinson) [reply docs/source/traits_user_manual/defining.rst]: Yes, that sounds good.

File: docs/source/traits_user_manual/defining.rst [hunk 5]
```diff
@@ -405,6 +405,56 @@ the table.
...
+
+.. _instance:
+
+Instance
+::::::::
+One of the most fundamental and useful predefined trait types is
+:class:`~.Instance`. Instance trait values are an instance of a particular class
+or its subclasses, as specified by the **klass** argument. **klass** can be
+either an instance of a class or a class itself (note this applies to all python
+classes, not necessarily just ``HasTraits`` subclasses).
```
Comment (mdickinson) [inline docs/source/traits_user_manual/defining.rst:418]: Suggest changing ``HasTraits`` to a reST link to the `HasTraits` class. (I know we're missing a lot of such links in existing docs, but I'd like to slowly move towards using them.)

```suggestion
classes, not necessarily just :class:`~.HasTraits` subclasses).
```

File: docs/source/traits_user_manual/defining.rst [hunk 6]
```diff
@@ -405,6 +405,56 @@ the table.
...
+
+If **klass** is an instance or if it is a class and **args** and **kw** are not
+specified, the default value is None. Otherwise, the default value is
```
Comment (mdickinson) [inline docs/source/traits_user_manual/defining.rst:421]:
```suggestion
specified, the default value is ``None``. Otherwise, the default value is
```

**Review by mdickinson — CHANGES_REQUESTED** [review]
Hi Aaron.

Thanks for the update.

This is a bit of a can of worms. A couple of things:

- I'm uncomfortable with documenting the `Instance(SomeClass())` pattern: I really don't want to do anything that would encourage its use. The right way to use `Instance` should always be to provide the type or interface that we want an instance of.
- Would it be possible to expand the documentation to cover the behaviour of the `allow_none` keyword? One gotcha that's worth noting is that even with `allow_none=False`, the default value of an `Instance` will still be `None`.

File: docs/source/traits_user_manual/defining.rst [hunk 7]
```diff
@@ -405,6 +405,58 @@ the table.
...
+
+Instance
+::::::::
+One of the most fundamental and useful predefined trait types is
+:class:`~.Instance`. Instance trait values are an instance of a particular class
+or its subclasses, as specified by the **klass** argument. **klass** can be
+either an instance of a class or a class itself (note this applies to all python
+classes, not necessarily just :class:`~.HasTraits` subclasses).  However, one should
+typically provide the type or interface they want an instance of, instead of
+providing an instance of a class already.
```
Comment (mdickinson) [inline docs/source/traits_user_manual/defining.rst:420]:
```suggestion
providing an instance of a class.
```

**Review by mdickinson — APPROVED** [review]
LGTM;  one minor suggested wording change.
